File: juno_rs/src/take_profit/basic.rs
use super::TakeProfit;
use crate::{genetics::Chromosome, Candle};
use juno_derive_rs::*;
use rand::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Chromosome, Clone, Debug, Deserialize, Serialize)]
pub struct BasicParams {
    pub up_threshold: f64,
    pub down_threshold: f64,
}

fn up_threshold(rng: &mut StdRng) -> f64 {
    rng.gen_range(0.001..10.000)
}
fn down_threshold(rng: &mut StdRng) -> f64 {
    up_threshold(rng)
}

pub struct Basic {
    up_threshold_factor: f64,
    down_threshold_factor: f64,
    close_at_position: f64,
    close: f64,
}

impl TakeProfit for Basic {
    type Params = BasicParams;

    fn new(params: &Self::Params) -> Self {
        Self {
            up_threshold_factor: 1.0 + params.up_threshold,
            down_threshold_factor: 1.0 - params.down_threshold,
            close_at_position: 0.0,
            close: 0.0,
        }
    }

    fn upside_hit(&self) -> bool {
        self.close >= self.close_at_position * self.up_threshold_factor
    }

    fn downside_hit(&self) -> bool {
        self.close <= self.close_at_position * self.down_threshold_factor
    }

    fn clear(&mut self, candle: &Candle) {
        self.close_at_position = candle.close;
    }

    fn update(&mut self, candle: &Candle) {
        self.close = candle.close;
    }
}

File: juno/take_profit/basic.py
from decimal import Decimal
from typing import Optional

from juno.candles import Candle

from .take_profit import TakeProfit


class Basic(TakeProfit):
    _up_threshold_factor: Decimal
    _down_threshold_factor: Decimal
    _close_at_position: Decimal = Decimal('0.0')
    _close: Decimal = Decimal('0.0')

    def __init__(self, up_threshold: Decimal, down_threshold: Optional[Decimal] = None) -> None:
        if down_threshold is None:
            down_threshold = up_threshold
        assert 0 <= up_threshold
        assert 0 <= down_threshold
        self._up_threshold_factor = 1 + up_threshold
        self._down_threshold_factor = 1 - down_threshold

    @property
    def upside_hit(self) -> bool:
        return self._close >= self._close_at_position * self._up_threshold_factor

    @property
    def downside_hit(self) -> bool:
        return self._close <= self._close_at_position * self._down_threshold_factor

    def clear(self, candle: Candle) -> None:
        self._close_at_position = candle.close

    def update(self, candle: Candle) -> None:
        self._close = candle.close
